### PR TITLE
Copter: fix take-off velocity limits

### PR DESCRIPTION
**Pull request by lthall**
This PR fixes problem where we are missing velocity limits that ensure maximums are not exceeded after position errors are created.

This is seen clearly when asking for a take-off with PILOT_TKOFF_ALT  set to a significant height.

### CODE REVIEW

File: libraries/AP_Math/control.cpp [hunk 1]
```diff
@@ -252,20 +248,20 @@ void shape_vel_accel_xy(const Vector2f& vel_input, const Vector2f& accel_input,
...
- The time constant must be positive.
+    maximum velocity - vel_min, vel_max,
+    maximum acceleration - accel_min, accel_max,
+    maximum jerk - jerk_max.
+ The jerk_max and accel_max must be positive. The accel_min must be negative.
  The function alters the variable accel to follow a jerk limited kinematic path to pos_input, vel_input and accel_input.
- The vel_max, vel_correction_max, and accel_max limits can be removed by setting the desired limit to zero.
+ The vel_max limit can be removed by setting the desired limit to zero.
+ The correction velocity can is limited to vel_max to vel_min. If limit_total_vel is true the total velocity is limited to vel_max to vel_min.
+ The correction acceleration can is limited to accel_max to accel_min. If limit_total_accel is true the total acceleration is limited to accel_max to accel_min.
```
Comment (lthall) [inline libraries/AP_Math/control.cpp:258]: Fix comment to remove: limit_total_accel

**Comment by tridge** [comment]
need to test on quadplanes, see if this negative impact on quadplane back transition

**Comment by lthall** [comment]
Before:
image
After:
image

**Review by IamPete1 — APPROVED** [review]
Looks good. I don't think this will affect quadplane because it is only used in position input methods `input_pos_xyz` and `input_pos_vel_accel_xy`. Quadplane does use them, but only in final landing not transition/approach stuff. Always worth double checking in the sim tho....

File: libraries/AP_Math/control.h [hunk 2]
```diff
@@ -51,12 +51,11 @@ void update_pos_vel_accel_xy(Vector2p& pos, Vector2f& vel, const Vector2f& accel
...
-    acceleration limits - accel_min, accel_max,
-    time constant - tc.
- The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
- The time constant also defines the time taken to achieve the maximum acceleration.
- The time constant must be positive.
+    maximum acceleration - accel_min, accel_max,
+    maximum jerk - jerk_max.
+ The jerk_max and accel_max must be positive. The accel_min must be negative.
  The function alters the variable accel to follow a jerk limited kinematic path to accel_input.
+ The correction acceleration can is limited to accel_max to accel_min. If limit_total_accel is true the total acceleration is limited to accel_max to accel_min.
```
[the remark this answers is not in the document]
Comment (lthall) [reply libraries/AP_Math/control.h:58]: yuck! Nice catch!

File: libraries/AP_Math/control.h [hunk 3]
```diff
@@ -71,14 +70,12 @@ void shape_accel_xy(const Vector3f& accel_input, Vector3f& accel,
...
- The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
- The time constant also defines the time taken to achieve the maximum acceleration.
- The time constant must be positive.
+    maximum velocity - vel_min, vel_max,
+    maximum acceleration - accel_min, accel_max,
+    maximum jerk - jerk_max.
+ The jerk_max and accel_max must be positive. The accel_min must be negative.
  The function alters the variable accel to follow a jerk limited kinematic path to vel_input and accel_input.
- The accel_max limit can be removed by setting it to zero.
+ The correction acceleration can is limited to accel_max to accel_min. If limit_total_accel is true the total acceleration is limited to accel_max to accel_min.
```
[the remark this answers is not in the document]
Comment (lthall) [reply libraries/AP_Math/control.h:78]: yuck! Nice catch!

File: libraries/AP_Math/control.cpp [hunk 4]
```diff
@@ -294,22 +290,27 @@ void shape_pos_vel_accel(postype_t pos_input, float vel_input, float accel_input
...
     if (is_negative(vel_min) && is_positive(vel_max)){
         vel_target = constrain_float(vel_target, vel_min, vel_max);
     }
 
     // velocity correction with input velocity
     vel_target += vel_input;
 
-    shape_vel_accel(vel_target, accel_input, vel, accel, accel_min, accel_max, jerk_max, dt, limit_total_accel);
+    // limit velocity between vel_min and vel_max
+    if (limit_total && is_negative(vel_min) && is_positive(vel_max)){
```
[the remark this answers is not in the document]
Comment (lthall) [reply libraries/AP_Math/control.cpp:302]: Thanks, I did this in a few places.

**Comment by davidbuzz** [comment]
@lthall offered beers to ppl that fix his comments, i heard him say it.  honest. :-)

File: libraries/AP_Math/control.cpp [hunk 5]
```diff
@@ -99,11 +99,9 @@ void update_pos_vel_accel_xy(Vector2p& pos, Vector2f& vel, const Vector2f& accel
 /* shape_accel calculates a jerk limited path from the current acceleration to an input acceleration.
  The function takes the current acceleration and calculates the required jerk limited adjustment to the acceleration for the next time dt.
  The kinematic path is constrained by :
-    acceleration limits - accel_min, accel_max,
-    time constant - tc.
- The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
- The time constant also defines the time taken to achieve the maximum acceleration.
- The time constant must be positive.
+    maximum acceleration - accel_min, accel_max,
```
Comment (tridge) [inline libraries/AP_Math/control.cpp:102]: there is no accel_min or accel_max in this function?

File: libraries/AP_Math/control.cpp [hunk 6]
```diff
@@ -143,14 +141,12 @@ void shape_accel_xy(const Vector3f& accel_input, Vector3f& accel,
 /* shape_vel_accel and shape_vel_xy calculate a jerk limited path from the current position, velocity and acceleration to an input velocity.
  The function takes the current position, velocity, and acceleration and calculates the required jerk limited adjustment to the acceleration for the next time dt.
  The kinematic path is constrained by :
-    maximum velocity - vel_max,
-    maximum acceleration - accel_max,
-    time constant - tc.
- The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
- The time constant also defines the time taken to achieve the maximum acceleration.
- The time constant must be positive.
+    maximum velocity - vel_min, vel_max,
```
Comment (tridge) [inline libraries/AP_Math/control.cpp:144]: there is no vel_min and vel_max

File: libraries/AP_Math/control.h [hunk 7]
```diff
@@ -51,12 +51,11 @@ void update_pos_vel_accel_xy(Vector2p& pos, Vector2f& vel, const Vector2f& accel
 /* shape_accel calculates a jerk limited path from the current acceleration to an input acceleration.
  The function takes the current acceleration and calculates the required jerk limited adjustment to the acceleration for the next time dt.
  The kinematic path is constrained by :
-    acceleration limits - accel_min, accel_max,
-    time constant - tc.
- The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
- The time constant also defines the time taken to achieve the maximum acceleration.
- The time constant must be positive.
+    maximum acceleration - accel_min, accel_max,
```
Comment (tridge) [inline libraries/AP_Math/control.h:54]: there is no accel_min and accel_max

File: libraries/AP_Math/control.h [hunk 8]
```diff
@@ -71,14 +70,12 @@ void shape_accel_xy(const Vector3f& accel_input, Vector3f& accel,
 /* shape_vel_accel and shape_vel_xy calculate a jerk limited path from the current position, velocity and acceleration to an input velocity.
  The function takes the current position, velocity, and acceleration and calculates the required jerk limited adjustment to the acceleration for the next time dt.
  The kinematic path is constrained by :
-    maximum velocity - vel_max,
-    maximum acceleration - accel_max,
-    time constant - tc.
- The time constant defines the acceleration error decay in the kinematic path as the system approaches constant acceleration.
- The time constant also defines the time taken to achieve the maximum acceleration.
- The time constant must be positive.
+    maximum velocity - vel_min, vel_max,
```
Comment (tridge) [inline libraries/AP_Math/control.h:73]: there is no vel_min and vel_max

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by IamPete1 — APPROVED** [review]
